destination_integration_test.go

```diff
@@ -15,89 +31,224 @@ import (
+	assertDeletedRecordIndex(ctx, t, is, index, id)
+
+	deleteAllRecords(is, index)
+}
+
+const maxRetries = 4
+
+func waitTime(i int) time.Duration {
+	wait := math.Pow(2, float64(i))
+	return time.Duration(wait) * time.Second
+}
+
+func assertWrittenRecordIndex(ctx context.Context, t *testing.T, is *is.I, index *pinecone.IndexConnection, id string, writtenVecs recordPayload) {
+	// Pinecone writes appear to be asynchronous. At the very least, in the current free tier serverless
+	// configuration that I've tested, pinecone writes occurred slightly after the RPC call
+	// returned data. Therefore, the following retry logic is needed to make tests more robust
+	for i := 1; i <= maxRetries; i++ {
+		res, err := index.FetchVectors(&ctx, []string{id})
+		is.NoErr(err)
+
+		vec, ok := res.Vectors[id]
+		if !ok {
+			if i == maxRetries {
+				is.Fail() // vector was not written
+			} else {
+				wait := waitTime(i)
+				t.Logf("retrying with wait of %v", wait)
+				time.Sleep(wait)
+				continue
+			}
+		}
+
+		is.Equal(vec.Values, writtenVecs.Values)
+		is.Equal(vec.SparseValues.Values, writtenVecs.SparseValues.Values)
+		is.Equal(vec.SparseValues.Indices, writtenVecs.SparseValues.Indices)
+		break
+	}
+}
+
+func assertDeletedRecordIndex(ctx context.Context, t *testing.T, is *is.I, index *pinecone.IndexConnection, id string) {
+	// same as assertWrittenRecordIndex, we need the retry for robustness
+	for i := 0; i <= maxRetries; i++ {
+		res, err := index.FetchVectors(&ctx, []string{id})
+		is.NoErr(err)
+
+		_, ok := res.Vectors[id]
+		if ok {
+			if i == maxRetries {
+				is.Fail() // vector found, not properly deleted
+			} else {
+				wait := waitTime(i)
+				t.Logf("retrying with wait of %v", wait)
+				time.Sleep(wait)
+				continue
+			}
+		}
+		break
+	}
+}
+
+func assertNamespaceExists(ctx context.Context, t *testing.T, is *is.I, index *pinecone.IndexConnection, namespace string) {
+	// same as assertWrittenRecordIndex, we need the retry for robustness
+	for i := 0; i <= maxRetries; i++ {
+		stats, err := index.DescribeIndexStats(&ctx)
+		is.NoErr(err)
+
+		_, namespaceExists := stats.Namespaces[namespace]
+		if !namespaceExists {
+			if i == maxRetries {
+				is.Fail() // vector found, not properly deleted
+			} else {
+				wait := waitTime(i)
+				t.Logf("retrying with wait of %v", wait)
+				time.Sleep(wait)
+				continue
+			}
+		}
+
+		break
+	}
+}
+
+func deleteAllRecords(is *is.I, index *pinecone.IndexConnection) {
+	ctx := context.Background()
+	err := index.DeleteAllVectorsInNamespace(&ctx)
+	is.NoErr(err)
 }
 
 func TestMain(t *testing.M) {
-	_ = godotenv.Load()
+	godotenv.Load()
+
 	t.Run()
 }
 
-func teardown(is *is.I, ctx context.Context, destination sdk.Destination) {
-	err := destination.Teardown(ctx)
+type connectorResource interface {
+	Teardown(ctx context.Context) error
+}
+
+func teardown(ctx context.Context, is *is.I, resource connectorResource) {
+	err := resource.Teardown(ctx)
 	is.NoErr(err)
```

Comment:
This can be a one-liner, `is.NoErr(resource.Teardown(ctx))`. Also, the resource is always a destination. With that, there are less reason to introduce a new interface and a new method for just the above.

Reply:
I remember I got the same comment before, sorry! I'll change it now.